packages/client/src/gfx/Board3D.tsx

```diff
@@ -69,14 +100,33 @@ export function Board3D(props: Props) {
         pointer.current.y = p.y;
     }, [pointer]);
 
-    const units = props.unitStates.map(u => 
-    (<Unit3D
-        key={u.id}
-        unit={u}
-        click={props.unitClick}
-        selected={props.selectedUnits.has(u.id)}
-        enemy={u.owner !== props.playerIndex}
-    />));
+    const units = props.unitStates.map(u => {
+        const catalogEntryFn = UNIT_DISPLAY_CATALOG[u.kind];
+        if (!catalogEntryFn)
+            throw new Error("No display catalog entry for unit" + u.kind);
+        const catalogEntry = catalogEntryFn();
+
+        if (catalogEntry.isBuilding) {

```

Comment:
Frankly this looks hideous, at least the props should be passed as an exploded object and not like this.